File: examples/print_client.py
#!/usr/bin/env python3
"""
Example Python client for ESP32 Thermal Printer Gateway
Demonstrates how to send print jobs via WiFi
"""

import requests
import sys
from datetime import datetime

# Gateway configuration
GATEWAY_IP = "192.168.4.1"  # Default AP mode IP
# GATEWAY_IP = "192.168.1.100"  # Example station mode IP

BASE_URL = f"http://{GATEWAY_IP}"


def check_status():
    """Check the gateway and printer status"""
    try:
        response = requests.get(f"{BASE_URL}/status", timeout=5)
        response.raise_for_status()
        status = response.json()
        
        print("Gateway Status:")
        print(f"  WiFi: {status['wifi']}")
        print(f"  Bluetooth: {'Connected' if status['btConnected'] else 'Disconnected'}")
        print(f"  Free Heap: {status['freeHeap']} bytes")
        print(f"  Uptime: {status['uptime']} seconds")
        
        return status['btConnected']
    except requests.exceptions.RequestException as e:
        print(f"Error checking status: {e}")
        return False


def print_text(text):
    """Send text to the printer"""
    try:
        response = requests.post(
            f"{BASE_URL}/print",
            data=text,
            headers={'Content-Type': 'text/plain'},
            timeout=10
        )
        response.raise_for_status()
        result = response.json()
        
        print(f"Print job sent successfully!")
        print(f"  Bytes sent: {result.get('bytes', 'unknown')}")
        return True
    except requests.exceptions.RequestException as e:
        print(f"Error sending print job: {e}")
        return False


def print_receipt_example():
    """Print a sample receipt"""
    receipt = f"""
{'='*32}
    EXAMPLE RECEIPT
{'='*32}

Date: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}

Items:
  Coffee           $2.50
  Croissant        $3.00
  Orange Juice     $2.00
              ----------
  Total:           $7.50

Thank you for your visit!

{'='*32}



"""
    return print_text(receipt)


def interactive_mode():
    """Interactive mode for testing"""
    print("\n=== ESP32 Thermal Printer Gateway Client ===\n")
    
    while True:
        print("\nOptions:")
        print("  1. Check status")
        print("  2. Print custom text")
        print("  3. Print sample receipt")
        print("  4. Exit")
        
        choice = input("\nEnter choice (1-4): ").strip()
        
        if choice == "1":
            check_status()
        elif choice == "2":
            print("\nEnter text to print (press Ctrl+D or Ctrl+Z when done):")
            try:
                lines = []
                while True:
                    line = input()
                    lines.append(line)
            except EOFError:
                text = "\n".join(lines) + "\n\n\n"
                print_text(text)
        elif choice == "3":
            print_receipt_example()
        elif choice == "4":
            print("Goodbye!")
            break
        else:
            print("Invalid choice. Please try again.")


def main():
    """Main function"""
    if len(sys.argv) > 1:
        # Command line argument provided
        if sys.argv[1] == "status":
            check_status()
        elif sys.argv[1] == "receipt":
            print_receipt_example()
        elif sys.argv[1] == "text":
            if len(sys.argv) > 2:
                text = " ".join(sys.argv[2:]) + "\n\n\n"
                print_text(text)
            else:
                print("Usage: python3 print_client.py text <message>")
        else:
            print(f"Usage: {sys.argv[0]} [status|receipt|text <message>]")
    else:
        # Interactive mode
        interactive_mode()


if __name__ == "__main__":
    main()

File: examples/print_client.js
/**
 * Node.js client example for ESP32 Thermal Printer Gateway
 * 
 * Install dependencies:
 *   npm install axios
 * 
 * Usage:
 *   node print_client.js status
 *   node print_client.js print "Hello World"
 */

const axios = require('axios');

// Gateway configuration
const GATEWAY_IP = process.env.GATEWAY_IP || '192.168.4.1';
const BASE_URL = `http://${GATEWAY_IP}`;

/**
 * Check gateway and printer status
 */
async function checkStatus() {
  try {
    const response = await axios.get(`${BASE_URL}/status`);
    const status = response.data;
    
    console.log('Gateway Status:');
    console.log(`  WiFi: ${status.wifi}`);
    console.log(`  Bluetooth: ${status.btConnected ? 'Connected' : 'Disconnected'}`);
    console.log(`  Free Heap: ${status.freeHeap} bytes`);
    console.log(`  Uptime: ${status.uptime} seconds`);
    
    return status.btConnected;
  } catch (error) {
    console.error('Error checking status:', error.message);
    return false;
  }
}

/**
 * Send text to the printer
 */
async function printText(text) {
  try {
    // Add extra line feeds for paper advance
    const printData = text + '\n\n\n';
    
    const response = await axios.post(
      `${BASE_URL}/print`,
      printData,
      {
        headers: { 'Content-Type': 'text/plain' }
      }
    );
    
    console.log('✓ Print job sent successfully!');
    console.log(`  Bytes sent: ${response.data.bytes || 'unknown'}`);
    return true;
  } catch (error) {
    console.error('✗ Error sending print job:', error.message);
    if (error.response) {
      console.error('  Response:', error.response.data);
    }
    return false;
  }
}

/**
 * Print a sample receipt
 */
async function printReceipt() {
  const now = new Date();
  const receipt = `
${'='.repeat(32)}
    EXAMPLE RECEIPT
${'='.repeat(32)}

Date: ${now.toLocaleString()}

Items:
  Coffee           $2.50
  Croissant        $3.00
  Orange Juice     $2.00
              ----------
  Total:           $7.50

Thank you for your visit!

${'='.repeat(32)}


`;
  
  return await printText(receipt);
}

/**
 * Main function
 */
async function main() {
  const args = process.argv.slice(2);
  
  if (args.length === 0) {
    console.log('Usage:');
    console.log('  node print_client.js status           - Check gateway status');
    console.log('  node print_client.js print <text>     - Print custom text');
    console.log('  node print_client.js receipt          - Print sample receipt');
    return;
  }
  
  const command = args[0];
  
  switch (command) {
    case 'status':
      await checkStatus();
      break;
      
    case 'print':
      if (args.length < 2) {
        console.error('Error: No text provided');
        console.log('Usage: node print_client.js print <text>');
        process.exit(1);
      }
      const text = args.slice(1).join(' ');
      await printText(text);
      break;
      
    case 'receipt':
      await printReceipt();
      break;
      
    default:
      console.error(`Unknown command: ${command}`);
      console.log('Valid commands: status, print, receipt');
      process.exit(1);
  }
}

// Run main function
main().catch(error => {
  console.error('Unexpected error:', error);
  process.exit(1);
});

module.exports = { checkStatus, printText, printReceipt };
